### ISSUE TITLE
Any interest in test.sh becoming test.py and using git submodules for update?

### ISSUE THREAD

**Issue by MattOates** [radeklat/python-before-push#1]
Hi Radek,

Im looking at porting 1:1 test.sh to core system Python 3 just so there is slightly more portability and ease of understanding given everything here is Python specific, it kind of makes sense to not leave Python. 

I'm also considering stripping most of the RC and update code, as that is a huge amount of what is happening in test.sh and replacing it with git submodule semantics. So test.py would update itself by finding a parent git repo and itself as a submodule then just moving that to the latest release tag. This avoids source commits in every code base using python-before-push and simpler exact versioning.

With respect to the testrc files which at the moment are raw bash I was considering using the INI format from configparser instead? https://docs.python.org/3/library/configparser.html#module-configparser

Any interest or thoughts?

**Comment by radeklat** [comment]
Hi Matt.

This script is trying to solve things that are already solved by other product, so I would suggest having a look at those first:

**Virtual environment and dependencies management**
Can be replaced with [pipenv](https://github.com/pypa/pipenv). The performance got much better from the last time I checked. Additional benefit is the it creates a reproducible environment, which test.sh doesn't.

**Hacking around testing tools**
Can be replaced with an existing Python automation framework. I would suggest [Invoke](http://www.pyinvoke.org/), because it is readable, minimal and has similar funtionality to Makefiles.

**Synchronisation**
You already mentioned you would like to use git sub-modules instead. Even if not, that could be very small, stand-alone script. Or maybe [cookiecutter](https://github.com/cookiecutter/cookiecutter) could be utilised for that.

**Makefile**
Although not used in this repository, I know that test.sh is often executed via Makefile. This could be as well replaced with and automation framework, such as Invoke. pipenv also has a simple facility for execution targets.

To sum up, it is probably not worth the time to write "another tool" when there are existing frameworks that could be used. I will likely rewrite test.sh into pipenv+Invoke in the future. Let me know if you would be interested in that as well.
